fetch.js

```diff
@@ -37,6 +37,25 @@
     return value
   }
 
+  function getIterator(headers, addItem) {
+    var items = []
+    headers.forEach(addItem, items)
+    var iterator = {
+      next: function() {
+        var value = items.shift()
+        return {done: value === undefined, value: value}
+      }
+    }
+
+    if (support.iterable) {
+      iterator[Symbol.iterator] = function() {
+        return iterator
+      }
+    }
+
+    return iterator
+  }
```

Comment:
We'll get most of the DRY benefit with less complexity if we reduce `getIterator`'s responsibilities to just creating and returning an iterator object over a list of values. It's currently also responsible for building the value list from a function parameter, which complicates each call site.


```js
  // Build a destructive iterator for the value list
  function iteratorFor(items) {
    var iterator = {
      next: function() {
        var value = items.shift()
        return {done: value === undefined, value: value}
      }
    }

    if (support.iterable) {
      iterator[Symbol.iterator] = function() {
        return iterator
      }
    }

    return iterator
  }

// Each call site still builds its own value list.
Headers.prototype.keys = function() {
  var items = []
  this.forEach(function(value, name) { items.push(name) })
  return iteratorFor(items)
}
```

I used `iteratorFor` as the function name in this example because a "get" prefix implies the function returns a saved, pre-calculated value. It's a factory function, though, so maybe `iteratorFor`, `iteratorFrom`, `createIterator`, or simply the noun `iterator` would be better names.